app/services/preserved_object_audit_service.rb

```diff
@@ -0,0 +1,218 @@
+# frozen_string_literal: true
+
+# A service for auditing complete_moabs against moabs on disk (and vice versa.)
+# If a preserved_object is missing, it is created.
+# If a complete_moab is missing for a moab, it is created.
+# For each complete_moab / moab, versions are compared and validations possibly performed.
+# Complete_moabs are deleted for any moved moabs (i.e., moab missing for complete_moab, but another moab exists in a different location)
+class PreservedObjectAuditService
+
+  attr_reader :druid, :all_results
+  attr_writer :logger
+
+  def initialize(druid)
+    @druid = druid
+    @logger = PreservationCatalog::Application.logger
+    # Each moab / complete moab has own audit result
+    create_all_results
+  end
+
+  def audit
+    unless preserved_object
+      # TODO: Is raising the correct way to handle?
+      raise "No preserved object and no moabs on disk for #{druid}" if moabs.empty?
+      create_preserved_object
+    end
+    raise 'preserved_object does not exist' unless preserved_object # Assertion
+
+    # For each moab, if there is no complete_moab then create one.
+    create_missing_complete_moabs
+
+    preserved_object.complete_moabs.all.each do |complete_moab|
+      moab = moab_for_complete_moab(complete_moab)
+
+      # Moab on disk does not exist
+      unless moab.exist?
+        handle_missing_moab(complete_moab)
+        next
+      end
+
+      if moab.current_version_id == complete_moab.version
+        handle_version_matches(complete_moab)
+        if complete_moab.status == 'ok'
+          complete_moab.update_audit_timestamps(false, true) # Version audited
+          next
+        end
+      end
+
+      next unless validate(moab, complete_moab) # Also adds to results and updates complete_moab
+
+      if moab.current_version_id < complete_moab.version
+        handle_unexpected_version(complete_moab)
+        next
+      end
+
+      update_status(complete_moab,'validity_unknown') # This will queue a checksum validation.
+
+      handle_bumped_version(complete_moab, moab) if moab.current_version_id > complete_moab.version
```

Comment:
Can handle_bumped_version result in a validity_unknown status? If so, do this before update_status, so we can trigger a CV on the new version.